mig-agent/agent.go

```diff
@@ -457,6 +484,182 @@ func startRoutines(ctx *Context) (err error) {
 		ctx.Channels.Log <- mig.Log{Desc: "periodic environment refresh is disabled"}
 	}
 
+	// check for persistent modules, if enabled parse their config options
+	if len(PERSISTENTMODULES) > 0 {
+		for _, moduleString := range PERSISTENTMODULES {
+			moduleName := moduleString[0]
+			ctx.Channels.Log <- mig.Log{Desc: fmt.Sprintf("Enabled persistent module %q", moduleName)}
+			//parse moduleparams from the config file
+			//module params are passed as JSON MsgClassParameters Message
+			moduleParams := moduleString[1]
+			ctx.Channels.Log <- mig.Log{Desc: fmt.Sprintf("Feeding message to persistent module %q", moduleParams)}
+			// feed params to peristent module
+			err = startPersistentModule(ctx, moduleName, moduleParams)
+			if err != nil {
+				return
+			}
+		}
+	}
+	return
+}
+
+// launch persistent module as a subprocess and open pipes to stdin and stdout
+// for communicating with the process
+// starts go routines that listens and feeds the subprocess
+func startPersistentModule(ctx *Context, moduleName, params string) (err error) {
+	var isTimeOut = true
+	//buffered channel may be ? design decision
+	stdOutChannel := make(chan []byte)
+	supervisorChan := make(chan persistentModuleMsg)
+	cmd := exec.Command(ctx.Agent.BinPath, "-m", moduleName)
+	// stdin, stdout are used for communicating with the persistent module process
+	// the flow is full duplex
+	// open pipes for communicating with process stdin and stdout
+	stdinPipe, err := cmd.StdinPipe()
+	if err != nil {
+		return err
+	}
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+	stdoutScanner := bufio.NewScanner(stdoutPipe)
```

Comment:
We are probably OK here, one thing we might want to consider though is there is a fixed length to lines the scanner will correctly parse, from the docs:

When a scan stops, the reader may have advanced arbitrarily far past the last token. Programs that need more control over error handling or large tokens, or must run sequential scans on a reader, should use bufio.Reader instead. 

If we run into issues with long messages coming back we might want to switch to this.